app/scripts/lib/signature/util.ts

```diff
@@ -0,0 +1,49 @@
+import type { SignatureController } from '@metamask/signature-controller';
+import type {
+  OriginalRequest,
+  TypedMessageParams,
+} from '@metamask/message-manager';
+import { endTrace, TraceName } from '../../../../shared/lib/trace';
+import { MESSAGE_TYPE } from '../../../../shared/constants/app';
+
+export type SignatureParams = [TypedMessageParams, OriginalRequest];
+
+export type MessageType = keyof typeof MESSAGE_TYPE;
+
+export type AddSignatureMessageRequest = {
+  signatureParams: SignatureParams;
+  signatureController: SignatureController;
+  type: MessageType;
+};
+
+const MESSAGE_TYPE_TO_FUNCTION = {
+  [MESSAGE_TYPE.ETH_SIGN_TYPED_DATA as MessageType]: 'newUnsignedTypedMessage',
+  [MESSAGE_TYPE.ETH_SIGN_TYPED_DATA_V1 as MessageType]:
+    'newUnsignedTypedMessage',
+  [MESSAGE_TYPE.ETH_SIGN_TYPED_DATA_V3 as MessageType]:
+    'newUnsignedTypedMessage',
+  [MESSAGE_TYPE.ETH_SIGN_TYPED_DATA_V4 as MessageType]:
+    'newUnsignedTypedMessage',
+  [MESSAGE_TYPE.PERSONAL_SIGN as MessageType]: 'newUnsignedPersonalMessage',
+} as Record<MessageType, keyof SignatureController>;
+
+export async function addSignatureMessage(request: AddSignatureMessageRequest) {
+  const { signatureParams, signatureController, type } = request;
+  const [_messageParams, signatureRequest] = signatureParams;
+  const { id } = signatureRequest;
+  const actionId = id?.toString();
+  const functionName = MESSAGE_TYPE_TO_FUNCTION[type] as keyof Pick<
+    SignatureController,
+    'newUnsignedTypedMessage' | 'newUnsignedPersonalMessage'
+  >;
+
+  endTrace({ name: TraceName.Middleware, id: actionId });
+
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore: Expected 4-5 arguments, but got 2.
+  const hash = await signatureController[functionName](...signatureParams);
```

Comment:
I still think it's a good idea to encapsulate the `SignatureController` usage in here, but should we still have separate exported functions for each message type to avoid all this reflection and casting?

Reply:
Done in the https://github.com/MetaMask/metamask-extension/pull/26967/commits/b5528fe838cd4be6e6fbaa0ed37ecb52ff66c2bf

Reply:
These functions are eventually bound as middleware (see https://github.com/MetaMask/metamask-extension/blob/34259a3c2a94a4669053c01993716c724c194ca1/app/scripts/lib/createMetamaskMiddleware.js#L29 ), so none of these functions are getting triggered manually by the UI - in other words we never had actions for them. 

I think this is looking very clean implementation that we intercept the `before` and `after` the controller action. @jpuri do you think we can add any comment to make this a bit more clear or readable?